From: Ben Greear <greearb@candelatech.com>

Only AP, adhoc and mesh mode needs to check CAC.
Stations, in particular, do not need this check.

Signed-off-by: Rubio Lu <Rubio-DW.Lu@mediatek.com>
Signed-off-by: Ben Greear <greearb@candelatech.com>
Signed-off-by: Chad Monroe <chad.monroe@smartrg.com>
---
v3:  Fix typo in SOB in 1/2, fix rebase typo in 2/2,
  split long line in 2/2
 .../net/wireless/mediatek/mt76/mt7615/mac.c   | 38 +++++++++++++++++--
 1 file changed, 35 insertions(+), 3 deletions(-)

--- a/mt7615/mac.c
+++ b/mt7615/mac.c
@@ -2196,6 +2196,32 @@ static int mt7615_dfs_start_radar_detect
 	return 0;
 }
 
+struct mt7615_vif_counts {
+	u32 mesh;
+	u32 adhoc;
+	u32 ap;
+};
+
+static void
+mt7615_vif_counts(void *priv, u8 *mac, struct ieee80211_vif *vif)
+{
+	struct mt7615_vif_counts *counts = priv;
+
+	switch (vif->type) {
+	case NL80211_IFTYPE_ADHOC:
+		counts->adhoc++;
+		break;
+	case NL80211_IFTYPE_MESH_POINT:
+		counts->mesh++;
+		break;
+	case NL80211_IFTYPE_AP:
+		counts->ap++;
+		break;
+	default:
+		break;
+	}
+}
+
 static int
 mt7615_dfs_init_radar_specs(struct mt7615_phy *phy)
 {
@@ -2239,6 +2265,7 @@ int mt7615_dfs_init_radar_detector(struc
 	struct mt7615_dev *dev = phy->dev;
 	bool ext_phy = phy != &dev->phy;
 	int err;
+	struct mt7615_vif_counts counts = {0};
 
 	if (is_mt7663(&dev->mt76))
 		return 0;
@@ -2266,9 +2293,14 @@ int mt7615_dfs_init_radar_detector(struc
 	phy->dfs_state = chandef->chan->dfs_state;
 
 	if (chandef->chan->flags & IEEE80211_CHAN_RADAR) {
-		if (chandef->chan->dfs_state != NL80211_DFS_AVAILABLE)
-			return mt7615_dfs_start_radar_detector(phy);
-
+		if (chandef->chan->dfs_state != NL80211_DFS_AVAILABLE) {
+			ieee80211_iterate_active_interfaces(phy->mt76->hw,
+				IEEE80211_IFACE_ITER_RESUME_ALL,
+				mt7615_vif_counts, &counts);
+			if (counts.ap + counts.adhoc + counts.mesh)
+				mt7615_dfs_start_radar_detector(phy);
+			return 0;
+		}
 		return mt7615_mcu_rdd_cmd(dev, RDD_CAC_END, ext_phy,
 					  MT_RX_SEL0, 0);
 	}
